ledger-service/http-json/src/main/scala/com/digitalasset/http/dbbackend/ContractDao.scala

```diff
@@ -30,36 +33,93 @@ import scalaz.syntax.order._
 import spray.json.{JsNull, JsValue}
 
 import java.io.{Closeable, IOException}
+import java.sql.SQLTransientConnectionException
 import java.util.concurrent.{ExecutorService, Executors, TimeUnit}
 import javax.sql.DataSource
 import scala.concurrent.ExecutionContext
 import scala.language.existentials
-import scala.util.Try
+import scala.util.{Failure, Success, Try}
+import scala.annotation.nowarn
 
 class ContractDao private (
     ds: DataSource with Closeable,
     xa: ConnectionPool.T,
     dbAccessPool: ExecutorService,
+    @nowarn diagnosticSetup: Option[
+      (DataSource with Closeable, ConnectionPool.T, ExecutorService)
+    ] = None,
+    @nowarn diagnosticConfig: Option[DiagnosticConfig] = None,
 )(implicit
     val jdbcDriver: SupportedJdbcDriver.TC
 ) extends Closeable {
 
   private val logger = LoggerFactory.getLogger(classOf[ContractDao])
   implicit val logHandler: log.LogHandler = Slf4jLogHandler(logger)
 
-  def transact[A](query: ConnectionIO[A]): IO[A] =
-    query.transact(xa)
+  def transact[A](query: ConnectionIO[A], clientInfo: Option[String] = None): IO[A] = {
+    val xa0 = (diagnosticConfig.flatMap(_.clientInfo), clientInfo) match {
+      case (Some(identifier), Some(info)) =>
+        Transactor.strategy.modify(
+          xa,
+          s => s.copy(before = s.before.flatMap(_ => setClientInfo(identifier, info))),
+        )
+      case _ => xa
+    }
+    diagnosticSetup match {
+      case Some((_, dxa, _)) =>
+        query.transact(xa0).handleErrorWith {
+          case e: SQLTransientConnectionException
+              if e.getMessage.contains("Connection is not available") =>
```

Comment:
In general yes. But to introduce additional conf option and variable to do throttling for code never to be used in production just for load-testing purposes is overkill I think.

Reply:
I don't think this feature is useable without throttling. If you're getting connection pool exhaustion, there's a good chance you're getting a lot of these at the same time, and hammering the diagnostic query potentially makes both the diagnosis experience and the underlying issue worse.

Reply:
Unless we are talking about using this feature in production (which is not intended) I don't see issue here. 
1. Result Set is dumped into logs as one log entry so there won't be any mixed log entries coming from different RS.
2. We already know performance characteristic from previous run when we found out that perf tests failed due to connection pool exhaustion. So this second run is to investigate connection pool issue. Even if it wasn't like so the impact on performance is negligible as the query is run by separate thread only in situation that connection pool is exhausted => very likely problem is on DB side rather than on application server side.

Look, I think it's good point, throttling is useful technique but here I think is just effort wasted. But if you want to I can implement it anyway no problem.